### PR TITLE
Reduce lock contention from Metrics Stats addition.

### LINKED ISSUE #1381: telemetry.Metric$Stat.add lock contention appears to cause a throughput bottleneck
Using `strest-grpc` to drive grpc unary traffic at varying levels of concurrency, I found some lock contention that may be contributing to #1379.

I collected lock contention with `curl "localhost:9990/admin/pprof/contention?seconds=100&hz=999" > locks_$concurrency` and then used `google-pprof` to inspect the output.

(For simplicity I'm going to chop off the bottom of each profile)

Concurrency: 10
```
Total: 2070 samples
    1870  90.3%  90.3%     1870  90.3% sun.nio.ch.EPollSelectorImpl.doSelect
     111   5.4%  95.7%      111   5.4% sun.nio.ch.EPollSelectorImpl.wakeup
       9   0.4%  96.1%        9   0.4% io.buoyant.telemetry.Metric$Stat.add
       7   0.3%  96.5%       10   0.5% com.twitter.util.Promise.continue
```

Concurrency: 20
```
(pprof) top
Total: 10183 samples
    9130  89.7%  89.7%     9130  89.7% sun.nio.ch.EPollSelectorImpl.doSelect
     681   6.7%  96.3%      681   6.7% sun.nio.ch.EPollSelectorImpl.wakeup
     107   1.1%  97.4%      107   1.1% io.buoyant.telemetry.Metric$Stat.add
      36   0.4%  97.8%       36   0.4% com.twitter.finagle.loadbalancer.LoadBalancerFactory$StackModule$$anon$2.status
      19   0.2%  97.9%       19   0.2% com.twitter.finagle.filter.NackAdmissionFilter.$anonfun$afterSend$1
```
Concurrency: 50
```
Total: 8663 samples
    4473  51.6%  51.6%     4473  51.6% sun.nio.ch.EPollSelectorImpl.doSelect
    1543  17.8%  69.4%     1543  17.8% sun.nio.ch.EPollSelectorImpl.wakeup
     987  11.4%  80.8%      987  11.4% io.buoyant.telemetry.Metric$Stat.add
     400   4.6%  85.5%      422   4.9% com.twitter.finagle.pool.CachingPool.apply
     374   4.3%  89.8%      374   4.3% com.twitter.finagle.loadbalancer.LoadBalancerFactory$StackModule$$anon$2.status
     174   2.0%  91.8%      232   2.7% com.twitter.finagle.pool.WatermarkPool$ServiceWrapper.close
     165   1.9%  93.7%      591   6.8% com.twitter.finagle.pool.WatermarkPool.apply
      70   0.8%  94.5%       70   0.8% com.twitter.finagle.filter.NackAdmissionFilter.$anonfun$afterSend$1
      58   0.7%  95.2%       58   0.7% com.twitter.finagle.util.Cache.put
```

Concurrency: 100
```
Total: 8101 samples
    1908  23.6%  23.6%     1908  23.6% io.buoyant.telemetry.Metric$Stat.add
    1312  16.2%  39.7%     1312  16.2% sun.nio.ch.EPollSelectorImpl.doSelect
    1081  13.3%  53.1%     1096  13.5% com.twitter.finagle.pool.CachingPool.apply
     857  10.6%  63.7%      857  10.6% com.twitter.finagle.loadbalancer.LoadBalancerFactory$StackModule$$anon$2.status
     702   8.7%  72.3%      702   8.7% sun.nio.ch.EPollSelectorImpl.wakeup
     535   6.6%  78.9%      640   7.9% com.twitter.finagle.pool.WatermarkPool$ServiceWrapper.close
     451   5.6%  84.5%     1553  19.2% com.twitter.finagle.pool.WatermarkPool.apply
     134   1.7%  86.2%      134   1.7% com.twitter.finagle.filter.NackAdmissionFilter.$anonfun$afterSend$1
     104   1.3%  87.4%      105   1.3% com.twitter.finagle.util.Cache.put
```

### PR DESCRIPTION
**Pull request by stevej**
Problem:
At high concurrency, Metrics$Stat.add becomes our top contention
hotspot, accounting for 25% of our thread blockage.

Solution:
Copy finagle's BucketedHistogram into our tree and use finer grained
locking to avoid contention.

Shortening the lock, specifically avoiding a binary search in the middle
of holding a lock reduces overall thread blocking by roughly 20% in a
load test with concurrency 100. This particular lock drops from the top
spot to number 6.

Testing:
The tests were copied from finagle.

Fixes #1381

### CODE REVIEW

File: telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    if (value >= Int.MaxValue) {
+      index = countsLength - 1
+      valueToAdd = Int.MaxValue
+    } else {
+      val asInt = value.toInt
+      // recall that limits represent upper bounds, exclusive — so take the next position (+1).
+      // we assume that no inputs can be larger than the largest value in the limits array.
+      index = Math.abs(util.Arrays.binarySearch(limits, asInt)) + 1
+      valueToAdd = value
+    }
```
Comment (olix0r) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: nit: i'd be inclined to write this as

```scala
val (index, valueToAdd) =
  if (value >= Int.MaxValue) (countsLength - 1, Int.MaxValue)
  else {
    ...
    (index, value)
  }
```

File: telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        // values greater than or equal to Int.MaxValue
+        val upperLimit = if (idx != limits.length) {
+          limits(idx)
+        } else Int.MaxValue
+        val lowerLimit = if (idx != 0) {
+          limits(idx - 1)
+        } else 0
+        BucketAndCount(lowerLimit, upperLimit, count)
+    }.toSeq
+  }
```
Comment (olix0r) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: eh, this is probably only called in metrics reporting so not high-priority

File: telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala (hunk 1 above)
Comment (stevej) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: Doesn't `(a,b)` allocate a `Tuple2` on the heap? That was why I avoided it.

File: telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala (hunk 2 above)
Comment (zackangelo) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: @olix0r do you know if 2.12 still performs a function allocation in cases like this?

**Comment by stevej** [comment]
Copied the tests over and I somehow broke the bucketing logic. `diff` leads me to believe that I didn't change anything in the logic.

**Comment by stevej** [comment]
Occam's razor strikes again: the bug was caused by my changes. This latest version simplifies the logic for easier merging back into finagle.

File: telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala (hunk 1 above)
Comment (stevej) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: I changed this to be a single var to make the diff from finagle smaller.

File: telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala (hunk 2 above)
Comment (stevej) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: Happy to make the change either way but it's not called that often, on the order of O(# of stats) per minute.

Comment (olix0r) [inline telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala]: @zackangelo unsure. that's a very good question

**Comment by adleong** [comment]
Going to close this for now until we have a benchmark that demonstrates that this improves things.

**Pull request closed without merging**